### PR TITLE
Add public API endpoint to redirect to current src-cli download.

### PR DESCRIPTION
**Pull request by efritz**
I gave #3231 a crack.

### CODE REVIEW

File: cmd/frontend/internal/httpapi/src_cli.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+func srcCliServe(w http.ResponseWriter, r *http.Request) error {
+	filename := mux.Vars(r)["rest"]
+	if !isExpectedRelease(filename) {
+		http.NotFound(w, r)
+		return nil
+	}
+
+	target := fmt.Sprintf("%s/%s/%s", srcCliDownloadsURL, SrcCliVersion, filename)
```
Comment (slimsag) [inline cmd/frontend/internal/httpapi/src_cli.go]: use `path.Join()` instead, which will properly handle URL escaping

Comment (slimsag) [inline cmd/frontend/internal/httpapi/src_cli.go]:
```suggestion
	target := path.Join(srcCliDownloadsURL, SrcCliVersion, filename)
```

File: cmd/frontend/internal/httpapi/src_cli_version.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package httpapi
+
+// Code generated by update_src_cli_version.go
+
+// SrcCliVersion is the latest src-cli release at the time of this build
+var SrcCliVersion = `3.9.0`
```
Comment (slimsag) [inline cmd/frontend/internal/httpapi/src_cli_version.go]: Should we not commit this file and just have it so it generates on CI before we build? Or will we update this file manually?

Comment (slimsag) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: Ah, I suppose as-is this would cause CI to fail if the version was ever incorrect because `go generate` would create a diff, I think?

**Comment by slimsag** [comment]
Super exciting! This will be great to have finally :) One suggestion inline, otherwise my main question here is around how the dev experience here is.

Can you also be sure to update docs in the src-cli repo if we need to take manual steps in this repo after cutting a release there?

File: cmd/frontend/internal/httpapi/src_cli_version.go (hunk 2 above)
Comment (efritz) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: This updates it forcefully instead of requiring user intervention. I could instead add a lint rule to ensure that it’s no different than what is checked in (unless there is some lint magic that makes this behavior already occur).

Comment (efritz) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: I’m not sure which way to lean, both experiences sound fine to me.

Comment (efritz) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: Ok, I see now that the `go-generate.sh` script will diff against all non-sum files after generation. This gives us a few options:

1) Remove the gen.go file and move it to a ./dev script that should be run manually after a src-cli version is cut. This has the disadvantage of being a manual step, but the advantage that our sg/sg builds will not fail if someone cuts an src-cli release.
2) Add this generated file to the ignore list. Have it be generated on each CI run. This should also require that the generated file be gitignored, otherwise it's this weird mutable thing that's in version control but doesn't reflect any real value of a build.
3) (_Not suggested for this point in time_) Find another way to determine valid versions (can also be future work). Right now we only have a minimum src-cli version, but no maximum (and since it's hard-coded into a value we can't update it post-release). Maybe we could have a public dot-com-only endpoint that maintains a range of versions for each sourcegraph release and all instances delegate to that. I'm not sure if there's additional implications here (if all deployments can reach dot-com without firewall rules).

Thoughts?

Comment (slimsag) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: Things I am OK with:

- Some manual steps required to release src-cli
- Some manual steps required in this repository when we release src-cli

Things I would not be happy about:

- Releasing src-cli breaks sg/sg CI (i.e. if there are manual steps in this repo to perform, they should be performed and merged *before* the CI here breaks as part of src-cli release steps)
- We can accidentally release a Sourcegraph version which references an outdated src-cli version.
- We cannot release a Sourcegraph patch version which continues to reference the src-cli version released with the original non-patch sourcegraph release, because CI fails or rejects it

Given those, do you have a best option in mind?

Comment (slimsag) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: Maybe for now it is as simple as not having a CI check in place and having the file manually generated, but very explicitly documented step in the src-cli release instructions?

Comment (efritz) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: Updates: The string isn't auto-generated and must be bumped by hand (I'll add docs to src-cli indicating this). The API will attempt to return any patch updates (but not major or minor updates) so that we can release fixes to older instances without requiring a patch release of sg itself.

How does that sound?

Comment (slimsag) [reply cmd/frontend/internal/httpapi/src_cli_version.go]: SGTM 👍

File: cmd/frontend/internal/httpapi/src_cli.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	"src_darwin_amd64",
+	"src_linux_amd64",
+	"src_windows_amd64.exe",
+}
+
+func srcCliVersionServe(w http.ResponseWriter, r *http.Request) error {
+	return writeJSON(w, &struct {
+		Version string `json:"version"`
+	}{
+		Version: srcCliVerison(),
```
Comment (slimsag) [inline cmd/frontend/internal/httpapi/src_cli.go]:
```suggestion
		Version: srcCliVersion(),
```

Why is Verizon here? 😛

File: internal/linkheader/extract.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package linkheader
```
Comment (slimsag) [inline internal/linkheader/extract.go]: Did you mean to bundle this change into this PR? it looks unrelated

Comment (slimsag) [reply internal/linkheader/extract.go]: nvm

File: internal/src-cli/consts.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package srccli
+
+// MinimumVersion is the minimum src-cli release version that works
+// with this instance. This must be updated manually between releases.
+// The public HTTP API will return this (or an updated patch version)
+// as the suggested download with this instance.
```
Comment (slimsag) [inline internal/src-cli/consts.go]:
```suggestion
// as the suggested download with this instance.
//
// At the time of a Sourcegraph release, this is always the latest src-cli version.
```

Comment (slimsag) [reply internal/src-cli/consts.go]: Is this the intent, or that this will almost never get bumped except when we make backwards-incompat changes to src-cli?

File: cmd/frontend/internal/httpapi/src_cli.go (hunk 3 above)
Comment (efritz) [reply cmd/frontend/internal/httpapi/src_cli.go]: Fixed.

File: internal/linkheader/extract.go (hunk 4 above)
Comment (efritz) [reply internal/linkheader/extract.go]: I moved a helper function that was defined privately in the lsifserver client so that it could be used here as well.

File: internal/src-cli/consts.go (hunk 5 above)
Comment (efritz) [reply internal/src-cli/consts.go]: (Sorry, missed this comment). This should be bumped on new non-patch features and on necessary breaking changes.

It will update patch versions automatically so that we can cherry pick into older src-cli tags if there is an issue with an older instance/src-cli pair.
